Map.cpp

```diff
@@ -0,0 +1,157 @@
+    current_path.push_back(0);
+    float current_path_sum = 0;
+    int current_position = 0;
+
+    while (current_path.size() < number_of_vertices) {
+        auto positions = get_highest_pheromone_list_from_vertex(current_position);
+        for (auto position: current_path) {
+            auto found_element = std::find(positions.begin(), positions.end(), position);
+            if (found_element != positions.end()) {
+                positions.erase(found_element);
+            }
+        }
+        int new_position = positions[0];
+        current_path.push_back(new_position);
+        current_path_sum += distances[current_position][new_position];
+        current_position = new_position;
+
+    }
+
+    current_path_sum += distances[current_path[number_of_vertices - 1]][0];
+
+
+    if (current_path_sum < best_path_sum) {
+        best_path_sum = current_path_sum;
+        best_path_vertices = current_path;
+    }
+}
+
+float Map::get_distance_on_path(int position_1, int position_2) {
+    return this->distances[position_1][position_2];
+}
+
+std::vector<int> Map::get_highest_pheromone_list_from_vertex(int position) {
+    std::vector<std::pair<int, float>> paths_list;
+    for (int i = 0; i < this->number_of_vertices; i++) {
+        if (i == position) {
+            continue;
+        }
+        std::pair<int, float> point = std::make_pair(i, this->pheromone_map[position][i]);
+        paths_list.push_back(point);
+    }
+    std::vector<int> vertices;
+
+    std::sort(paths_list.begin(), paths_list.end(),
+              [](const std::pair<int, float> &a, const std::pair<int, float> &b) {
+                  return a.second > b.second;
+              });
+
+    for (auto point: paths_list) {
+        vertices.push_back(point.first);
+    }
+
+    return vertices;
+}
+
+std::vector<int> Map::get_best_path_vertices() {
+    return this->best_path_vertices;
+}
+
+float Map::load_map(const std::string &path) {
+    std::ifstream file;
+    file.open(path, std::ios::in);
+
+    std::string filename;
+    float min_distance = inf;
+    if (file) {
+        file >> filename;
+        file >> this->number_of_vertices;
+        this->distances = std::vector(this->number_of_vertices, std::vector<float>(this->number_of_vertices, 0));
+
+        for (int i = 0; i < this->number_of_vertices; i++) {
+            for (int ii = 0; ii < this->number_of_vertices; ii++) {
+                float distance;
+                file >> distance;
+                this->distances[i][ii] = distance;
+                if (distance < min_distance && distance > 0) {
+                    min_distance = distance;
+                }
+            }
+        }
+        file >> this->global_best_path_sum;
+    } else {
+        // exception
+        std::cout << "No map found" << std::endl;
+    }
+
+    file.close();
+    return min_distance;
+}
+
+int Map::get_best_path_sum() {
+    return this->best_path_sum;
+}
+
+Map::Map() {}// Why do i need this???
```

Comment:
this is needed because the Map field in the Simulator class needs to be default constructed, this can be avoided by initializing the map field in the Simulator ctor initializer list